File: library/src/main/java/cc/shinichi/library/view/helper/DragCloseView.java
package cc.shinichi.library.view.helper;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.LinearInterpolator;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.devbrackets.android.exomedia.ui.widget.VideoView;

import cc.shinichi.library.ImagePreview;
import cc.shinichi.library.view.nine.ViewHelper;
import cc.shinichi.library.view.photoview.PhotoView;
import cc.shinichi.library.view.subsampling.SubsamplingScaleImageView;

/**
 * @author 工藤
 * @email qinglingou@gmail.com
 * create at 2018/10/19  11:22
 * description:辅助下拉关闭图片
 */
public class DragCloseView extends RelativeLayout {

    private static final String TAG = DragCloseView.class.getSimpleName();

    private final static int MAX_EXIT_Y = 500;
    private final static long DURATION = 200;

    private SubsamplingScaleImageView imageStatic;
    private PhotoView imageAnime;
    private VideoView videoView;

    private float mDownX;
    private float mDownY;
    private float mTranslationY;

    private onAlphaChangedListener mOnAlphaChangedListener;

    public DragCloseView(Context context) {
        this(context, null);
    }

    public DragCloseView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DragCloseView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        imageStatic = (SubsamplingScaleImageView) getChildAt(0);
        imageAnime = (PhotoView) getChildAt(1);
        videoView = (VideoView) getChildAt(2);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        if (ImagePreview.Companion.getInstance().isEnableDragClose()) {
            int action = ev.getActionMasked();
            if (action == MotionEvent.ACTION_DOWN) {
                mDownX = ev.getRawX();
                mDownY = ev.getRawY();
            } else if (action == MotionEvent.ACTION_MOVE) {
                return canInterceptDrag(ev);
            }
        }
        return false;
    }

    private boolean canInterceptDrag(MotionEvent ev) {
        float diffX = Math.abs(ev.getX() - mDownX);
        float diffY = Math.abs(ev.getY() - mDownY);
        if (diffY <= diffX) {
            // 下拉手势，Y 轴位移小于 X 轴位移，是横向滑动，不拦截
            return false;
        }
        getParent().requestDisallowInterceptTouchEvent(true);
        if (imageStatic != null && imageStatic.getVisibility() == View.VISIBLE) {
            // 静图
            boolean isAtEdge = ImagePreview.Companion.getInstance().isEnableDragCloseIgnoreScale()
                    ? imageStatic.getScale() <= (imageStatic.getMinScale() + 0.001F) || imageStatic.isAtYEdge()
                    : imageStatic.getScale() <= (imageStatic.getMinScale() + 0.001F) && imageStatic.isAtYEdge();
            return isAtEdge && (imageStatic.getMaxTouchCount() == 0 || imageStatic.getMaxTouchCount() == 1);
        } else if (imageAnime != null && imageAnime.getVisibility() == View.VISIBLE) {
            // 动图
            return imageAnime.getScale() <= (imageAnime.getMinimumScale() + 0.001F) &&
                    (imageAnime.getMaxTouchCount() == 0 || imageAnime.getMaxTouchCount() == 1);
        } else if (videoView != null && videoView.getVisibility() == View.VISIBLE) {
            // 视频
            return true;
        }
        return false;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int action = event.getAction() & event.getActionMasked();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                mDownX = event.getRawX();
                mDownY = event.getRawY();
            case MotionEvent.ACTION_MOVE:
                if (ImagePreview.Companion.getInstance().isEnableDragClose()) {
                    if ((imageAnime != null && imageAnime.getVisibility() == View.VISIBLE)
                            || (imageStatic != null && imageStatic.getVisibility() == View.VISIBLE)
                            || (videoView != null && videoView.getVisibility() == View.VISIBLE)
                    ) {
                        onOneFingerPanActionMove(event);
                    }
                }
                break;
            case MotionEvent.ACTION_UP:
                onActionUp();
                break;
            default:
                break;
        }
        return true;
    }

    /**
     * 处理拖拽事件
     *
     * @param event
     */
    private void onOneFingerPanActionMove(MotionEvent event) {
        float moveY = event.getRawY();
        mTranslationY = moveY - mDownY;
        if (mOnAlphaChangedListener != null) {
            mOnAlphaChangedListener.onTranslationYChanged(event, mTranslationY);
        }
        setTranslationY(mTranslationY);
    }

    private void onActionUp() {
        // 是否启用上拉关闭
        boolean enableUpDragClose = ImagePreview.Companion.getInstance().isEnableUpDragClose();
        if (enableUpDragClose) {
            if (Math.abs(mTranslationY) > MAX_EXIT_Y) {
                exit(mTranslationY);
            } else {
                resetCallBackAnimation();
            }
        } else {
            if (mTranslationY > MAX_EXIT_Y) {
                exit(mTranslationY);
            } else {
                resetCallBackAnimation();
            }
        }
    }

    private void resetCallBackAnimation() {
        ValueAnimator animatorY = ValueAnimator.ofFloat(mTranslationY, 0);
        animatorY.setDuration(DURATION);
        animatorY.addUpdateListener(animation -> {
            mTranslationY = (float) animation.getAnimatedValue();
            setTranslationY(mTranslationY);
        });
        animatorY.addListener(new SimpleAnimatorListener() {
            @Override
            public void onAnimationEnd(@NonNull Animator animation) {
                mTranslationY = 0;
                reset();
            }
        });
        animatorY.start();
    }

    private void exit(float currentY) {
        // 默认的退出动画
        ValueAnimator animatorExit;
        if (currentY > 0) {
            animatorExit = ValueAnimator.ofFloat(mTranslationY, getHeight());
        } else {
            animatorExit = ValueAnimator.ofFloat(mTranslationY, -getHeight());
        }
        animatorExit.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(@NonNull ValueAnimator animation) {
                float fraction = (float) animation.getAnimatedValue();
                ViewHelper.setScrollY(DragCloseView.this, -(int) fraction);
            }
        });
        animatorExit.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(@NonNull Animator animation) {
            }

            @Override
            public void onAnimationEnd(@NonNull Animator animation) {
                reset();
                ((Activity) getContext()).finish();
            }

            @Override
            public void onAnimationCancel(@NonNull Animator animation) {
            }

            @Override
            public void onAnimationRepeat(@NonNull Animator animation) {
            }
        });
        animatorExit.setDuration(DURATION);
        animatorExit.setInterpolator(new LinearInterpolator());
        animatorExit.start();
    }

    /**
     * 暴露的回调方法（可根据位移距离或者alpha来改变主UI控件的透明度等
     */
    public void setOnAlphaChangeListener(onAlphaChangedListener alphaChangeListener) {
        mOnAlphaChangedListener = alphaChangeListener;
    }

    private void reset() {
        if (null != mOnAlphaChangedListener) {
            mOnAlphaChangedListener.onTranslationYChanged(null, mTranslationY);
        }
    }

    public abstract static class SimpleAnimatorListener implements Animator.AnimatorListener {
        @Override
        public void onAnimationStart(@NonNull Animator animation) {
        }

        @Override
        public void onAnimationCancel(@NonNull Animator animation) {
        }

        @Override
        public void onAnimationRepeat(@NonNull Animator animation) {
        }
    }

    public interface onAlphaChangedListener {

        void onTranslationYChanged(MotionEvent event, float translationY);
    }
}

File: library/src/main/java/cc/shinichi/library/ui/widget/DragCloseView.kt
package cc.shinichi.library.ui.widget

import android.animation.Animator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.animation.LinearInterpolator
import android.widget.RelativeLayout
import cc.shinichi.library.ImagePreview
import cc.shinichi.library.ui.photoview.PhotoView
import cc.shinichi.library.ui.subsampling.SubsamplingScaleImageView
import cc.shinichi.library.ui.widget.nine.ViewHelper
import kotlin.math.abs

/**
 * 拖拽关闭辅助 View
 *
 * 支持上下拖拽关闭图片预览
 *
 * @author 工藤
 * @email qinglingou@gmail.com
 */
class DragCloseView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {

    companion object {
        private const val TAG = "DragCloseView"
        private const val MAX_EXIT_Y = 500
        private const val DURATION = 200L
    }

    private var imageStatic: SubsamplingScaleImageView? = null
    private var imageAnime: PhotoView? = null
    private var videoView: View? = null

    private var downX = 0f
    private var downY = 0f
    private var translationYValue = 0f

    private var alphaChangedListener: OnAlphaChangedListener? = null

    override fun onFinishInflate() {
        super.onFinishInflate()
        imageStatic = getChildAt(0) as? SubsamplingScaleImageView
        imageAnime = getChildAt(1) as? PhotoView
        videoView = getChildAt(2)
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        if (!ImagePreview.instance.isEnableDragClose) {
            return false
        }

        return when (ev.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                downX = ev.rawX
                downY = ev.rawY
                false
            }

            MotionEvent.ACTION_MOVE -> canInterceptDrag(ev)
            else -> false
        }
    }

    private fun canInterceptDrag(ev: MotionEvent): Boolean {
        val diffX = abs(ev.x - downX)
        val diffY = abs(ev.y - downY)

        // Y 轴位移小于 X 轴位移，是横向滑动，不拦截
        if (diffY <= diffX) {
            return false
        }

        parent.requestDisallowInterceptTouchEvent(true)

        return when {
            // 静图
            imageStatic?.visibility == View.VISIBLE -> {
                val isAtEdge = if (ImagePreview.instance.isEnableDragCloseIgnoreScale) {
                    imageStatic!!.scale <= imageStatic!!.minScale + 0.001f || imageStatic!!.isAtYEdge
                } else {
                    imageStatic!!.scale <= imageStatic!!.minScale + 0.001f && imageStatic!!.isAtYEdge
                }
                isAtEdge && (imageStatic!!.maxTouchCount == 0 || imageStatic!!.maxTouchCount == 1)
            }
            // 动图
            imageAnime?.visibility == View.VISIBLE -> {
                imageAnime!!.scale <= imageAnime!!.minimumScale + 0.001f &&
                        (imageAnime!!.maxTouchCount == 0 || imageAnime!!.maxTouchCount == 1)
            }
            // 视频
            videoView?.visibility == View.VISIBLE -> true
            else -> false
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action and event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                downX = event.rawX
                downY = event.rawY
            }

            MotionEvent.ACTION_MOVE -> {
                if (ImagePreview.instance.isEnableDragClose) {
                    val hasVisibleContent = listOf(imageAnime, imageStatic, videoView)
                        .any { it?.visibility == View.VISIBLE }
                    if (hasVisibleContent) {
                        onOneFingerPanActionMove(event)
                    }
                }
            }

            MotionEvent.ACTION_UP -> onActionUp()
        }
        return true
    }

    /**
     * 处理拖拽事件
     */
    private fun onOneFingerPanActionMove(event: MotionEvent) {
        translationYValue = event.rawY - downY
        alphaChangedListener?.onTranslationYChanged(event, translationYValue)
        translationY = translationYValue
    }

    private fun onActionUp() {
        val enableUpDragClose = ImagePreview.instance.isEnableUpDragClose

        val shouldExit = if (enableUpDragClose) {
            abs(translationYValue) > MAX_EXIT_Y
        } else {
            translationYValue > MAX_EXIT_Y
        }

        if (shouldExit) {
            alphaChangedListener?.onExit()
            exit()
        } else {
            resetCallBackAnimation()
        }
    }

    private fun resetCallBackAnimation() {
        ValueAnimator.ofFloat(translationYValue, 0f).apply {
            duration = DURATION
            addUpdateListener { animation ->
                translationYValue = animation.animatedValue as Float
                translationY = translationYValue
            }
            addListener(object : SimpleAnimatorListener() {
                override fun onAnimationEnd(animation: Animator) {
                    translationYValue = 0f
                    reset()
                }
            })
            start()
        }
    }

    private fun exit() {
        val targetY = if (translationYValue > 0) height.toFloat() else -height.toFloat()

        ValueAnimator.ofFloat(translationYValue, targetY).apply {
            addUpdateListener { animation ->
                val fraction = animation.animatedValue as Float
                ViewHelper.setScrollY(this@DragCloseView, -fraction.toInt())
            }
            addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animation: Animator) {}
                override fun onAnimationEnd(animation: Animator) {
                    (context as? Activity)?.finish()
                }

                override fun onAnimationCancel(animation: Animator) {}
                override fun onAnimationRepeat(animation: Animator) {}
            })
            duration = DURATION
            interpolator = LinearInterpolator()
            start()
        }
    }

    /**
     * 设置透明度变化监听器
     */
    fun setOnAlphaChangeListener(listener: OnAlphaChangedListener?) {
        alphaChangedListener = listener
    }

    private fun reset() {
        alphaChangedListener?.onTranslationYChanged(null, translationYValue)
    }

    /**
     * 透明度变化监听接口
     */
    interface OnAlphaChangedListener {
        /**
         * Y 轴位移变化回调
         */
        fun onTranslationYChanged(event: MotionEvent?, translationY: Float)

        /**
         * 退出回调
         */
        fun onExit()
    }

    /**
     * 简化的动画监听器
     */
    abstract class SimpleAnimatorListener : Animator.AnimatorListener {
        override fun onAnimationStart(animation: Animator) {}
        override fun onAnimationCancel(animation: Animator) {}
        override fun onAnimationRepeat(animation: Animator) {}
    }
}
